sa/sa.go

```diff
@@ -835,6 +835,12 @@ func addRevokedCertificate(ctx context.Context, tx db.Executor, req *sapb.Revoke
 		NotAfterHour: serial.Expires.Add(time.Hour).Truncate(time.Hour),
 	})
 	if err != nil {
+		if db.IsDuplicate(err) {
+			// An attempted duplicate insert means that this certificate was already
+			// revoked. The RA has special logic for that case, so use the specific
+			// error for it.
+			return berrors.AlreadyRevokedError("certificate with serial %s already in revokedCertificates table", req.Serial)
+		}
```

Comment:
This will never happen because there is no index on `serial` in this table:

https://github.com/letsencrypt/boulder/blob/d9f4bfc258695c98d511635b68981c13764e0821/sa/db/boulder_sa/20230919000000_RevokedCertificates.sql#L12-L13

Reply:
Good news: we're adding that index for unrelated reasons: https://github.com/letsencrypt/boulder/pull/8427
Bad news: it won't be a unique key index, so I'll still need to update this PR with some select-and-insert-in-a-transaction shenanigans